src/stan/mcmc/chainset.hpp

```diff
@@ -0,0 +1,510 @@
+   * diagnostic for a set of per-chain draws, for bulk and tail Rhat.
+   * Based on paper https://arxiv.org/abs/1903.08008
+   *
+   * @param name parameter name
+   * @return pair (bulk_rhat, tail_rhat)
+   */
+  std::pair<double, double> split_rank_normalized_rhat(
+      const std::string& name) const {
+    return split_rank_normalized_rhat(index(name));
+  }
+
+  /**
+   * Computes the effective sample size (ESS) for the specified
+   * parameter across all chains, according to the algorithm presented in
+   * https://arxiv.org/abs/1903.08008, section 3.2 for folded (split)
+   * rank-normalized ESS for both bulk and tail ESS.
+   *
+   * @param index parameter index
+   * @return pair (bulk_ess, tail_ess)
+   */
+  std::pair<double, double> split_rank_normalized_ess(const int index) const {
+    return analyze::split_rank_normalized_ess(samples(index));
+  }
+
+  /**
+   * Computes the effective sample size (ESS) for the specified
+   * parameter across all chains, according to the algorithm presented in
+   * https://arxiv.org/abs/1903.08008, section 3.2 for folded (split)
+   * rank-normalized ESS for both bulk and tail ESS.
+   *
+   * @param name parameter name
+   * @return pair (bulk_ess, tail_ess)
+   */
+  std::pair<double, double> split_rank_normalized_ess(
+      const std::string& name) const {
+    return split_rank_normalized_ess(index(name));
+  }
+
+  /**
+   * Computes the mean Monte Carlo error estimate for the central 90% interval.
+   * See https://arxiv.org/abs/1903.08008, section 4.4.
+   * Follows implementation in the R posterior package
+   *
+   * @param index parameter index
+   * @return mcse
+   */
+  double mcse_mean(const int index) const {
+    if (num_samples() < 4
+        || !stan::analyze::is_finite_and_varies(samples(index)))
+      return std::numeric_limits<double>::quiet_NaN();
+    double ess = analyze::ess(samples(index));
+    return sd(index) / std::sqrt(ess);
+  }
+
+  /**
+   * Computes the mean Monte Carlo error estimate for the central 90% interval.
+   * See https://arxiv.org/abs/1903.08008, section 4.4.
+   * Follows implementation in the R posterior package.
+   *
+   * @param name parameter name
+   * @return mcse
+   */
+  double mcse_mean(const std::string& name) const {
+    return mcse_mean(index(name));
+  }
+
+  /**
+   * Computes the standard deviation of the Monte Carlo error estimate
+   * https://arxiv.org/abs/1903.08008, section 4.4.
+   * Follows implementation in the R posterior package.
+   *
+   * @param index parameter index
+   * @return mcse_sd
+   */
+  double mcse_sd(const int index) const {
+    if (num_samples() < 4
+        || !stan::analyze::is_finite_and_varies(samples(index)))
+      return std::numeric_limits<double>::quiet_NaN();
+    Eigen::MatrixXd s = samples(index);
+    Eigen::MatrixXd s2 = s.array().square();
+    double ess_s = analyze::ess(s);
+    double ess_s2 = analyze::ess(s2);
+    double ess_sd = std::min(ess_s, ess_s2);
```

Comment:
I can't find a matching implementation in posterior. The one I see doesn't do this min calculation:

https://github.com/stan-dev/posterior/blob/90a07cce51fdde484e97596cdd186adf265d8f5e/R/convergence.R#L490-L506

I think it would be nice if MCSE for the mean and for the standard deviation both lived in their own files, like how the ess/rhat functions do, and this class just called them. That would allow for expanded documentation and testing

Reply:
I cannot figure out why this is written this way - perhaps I'm the person who copied it from somewhere else - mea culpa, but seriously, this looks quite wrong.  replacing with implementation in posterior, written by Aki.